pmpro-member-directory.php

```diff
@@ -158,3 +158,103 @@ function pmpromd_plugin_row_meta($links, $file) {
 	return $links;
 }
 add_filter('plugin_row_meta', 'pmpromd_plugin_row_meta', 10, 2);
+
+/**
+ * We determine that the URL base is for the profile and then set up the rewrite rule
+ */
+function pmpromd_custom_rewrite_rules() {
+
+	global $pmpro_pages;
+
+	$parent_id = get_post_field( 'post_parent', $pmpro_pages['profile'] );
+	$slug = get_post_field( 'post_name',  $pmpro_pages['profile'] );
+
+	if( !empty( $parent_id ) ){
+		$parent_slug = get_post_field( 'post_name',  $parent_id );
+		$profile_base = $parent_slug.'/'.$slug;
+	} else {
+		$profile_base = $slug;
+	}
+
+	add_rewrite_rule(
+		$profile_base.'/([^/]+)/?$',
+		'index.php?pagename='.$profile_base.'&pu=$matches[1]',
+		'top'
+	);
+
+}
+add_action('init', 'pmpromd_custom_rewrite_rules', 10 );
+
+/**
+ * Adding in the ?pu parameter so that we can retrieve the value from the pretty permalink
+ */
+function pmpromd_custom_query_vars( $query_vars ) {
+
+    $query_vars[] = 'pu';
+    
+    return $query_vars;
+}
+
+add_filter( 'query_vars', 'pmpromd_custom_query_vars', 10, 1 );
+
+/**
+ * On Page Settings Save, Flush Rewrite Rules
+ */
+function pmpromd_pagesettings_flush(){
+
+	if( 
+		!empty( $_REQUEST['page'] ) && 
+		$_REQUEST['page'] == 'pmpro-pagesettings' && //Are we on the PMPro Page Settings
+		!empty( $_REQUEST['savesettings']) && //Are we hitting the save button
+		( !empty( $_REQUEST['profile_page_id'] ) ) //Is there a profile page present
+	){
+
+		flush_rewrite_rules( true );
+	}
+
+}
+add_action( 'admin_init', 'pmpromd_pagesettings_flush' );
+
+/**
+ * We're saving a page, is it a Profile page
+ */
+function pmpromd_page_save_flush( $post_id ){
+
+	global $pmpro_pages;
+
+	if( !empty( $pmpro_pages['profile'] ) ){
+		if( (int)$pmpro_pages['profile'] == $post_id ){
+			flush_rewrite_rules( true );
+		}
+	}
+
+}
+add_action( 'save_post', 'pmpromd_page_save_flush', 10, 1 );
+
+/**
+ * Flushing rewrite rules on plugin activate as a precautionary measure
+ */
+function pmpromd_flush_on_activate(){
+
+	flush_rewrite_rules( true );
+	
+}
+register_activation_hook( __FILE__, 'pmpromd_flush_on_activate' );
+
+/**
+ * Redirect from the OLD profile URL to the new URL's
+ */
+function pmpromd_redirect_profile_links(){
+
+	if( !empty( $_REQUEST['pu'] ) ){
+
+		global $pmpro_pages;
+
+		$profile_url = get_the_permalink( $pmpro_pages['profile'] );
+
+		wp_redirect( $profile_url.'/'.$_REQUEST['pu'], 302, 'WordPress' );
```

Comment:
This looks like trouble, we should sanitize it with `sanitize_text_field()` or something to ensure it's URL-safe.

Reply:
We should make a function that builds the profile link using the profile URL (in the function) and the $path as. an argument for the path to use.

It should ensure that the path is URL-safe, perhaps `sanitize_title()` would be good there.